// client/src/features/shared/ui/fields/InputField.tsx
import {
  useState,
  type HTMLInputAutoCompleteAttribute,
  type HTMLInputTypeAttribute,
} from "react"
import visibilityOnIcon from "@icons/misc/visibility-on.svg"
import visibilityOffIcon from "@icons/misc/visibility-off.svg"
import { IconCheckbox } from "@features/shared"

interface Props extends Pick<React.ComponentProps<"div">, "id" | "className"> {
  name?: string | undefined
  value?: string | readonly string[] | number | undefined
  onValueChanged?: (newValue: string) => void
  type?: HTMLInputTypeAttribute | undefined
  placeholder?: string | undefined
  autocomplete?: HTMLInputAutoCompleteAttribute | undefined
  hideable?: boolean
  isHidden?: boolean
  extendWidth?: boolean
  indicateError?: boolean
}
const InputField = ({
  id,
  name,
  value = undefined,
  onValueChanged,
  type = "text",
  placeholder = "Input",
  autocomplete,
  className = "",
  hideable,
  isHidden,
  extendWidth,
  indicateError,
}: Props) => {
  const [text, setText] = useState(value || "")
  const [hidden, setHidden] = useState(hideable && isHidden)

  const onChangeCallback = (event: React.ChangeEvent<HTMLInputElement>) => {
    const newValue = event.target.value

    setText(newValue)

    if (onValueChanged) {
      onValueChanged(newValue)
    }
  }
  const expand = !!text

  return (
    <div
      className={`relative bg-stone-950 flex h-12 items-center px-3 py-2 transition-[background-color,height] hover:bg-stone-900 focus:border focus:border-stone-300 
        ${expand && "h-16"} ${indicateError && "border border-red-900"} ${
        extendWidth && "w-full"
      }`}
    >
      <label
        htmlFor={id}
        className={`absolute left-3 pointer-events-none transition-all ${
          expand
            ? "top-1.5 text-[0.85rem] text-stone-600 translate-y-0"
            : "top-1/2 text-[1rem] text-stone-500 -translate-y-1/2"
        }`}
      >
        {placeholder}
      </label>
      <input
        id={id}
        name={name}
        value={text}
        onChange={onChangeCallback}
        type={hidden ? "password" : type}
        autoComplete={autocomplete}
        className={`bg-transparent border-0 outline-0 h-auto text-stone-200 p-0 m-0 ${
          expand && "pt-5"
        } ${className} ${extendWidth && "w-100"}`}
      />
      {hideable && (
        <IconCheckbox
          value={hidden}
          onValueChanged={setHidden}
          checkedIcon={visibilityOffIcon}
          uncheckedIcon={visibilityOnIcon}
        />
      )}
    </div>
  )
}
export default InputField
